README.md

```diff
@@ -1,2 +1,80 @@
-# beacon-starter
+# Beacon Reader Example
 
+> A starter project for reading beacon values from a smart contract
+
+This project is composed of the following:
+
+1. A test that uses a mocked RrpBeaconServer to simulate reading a beacon value from a smart contract.
+1. A script to deploy the BeaconReaderExample contract to any network. This script will set the address of the RrpBeaconServer contract deployed on the selected network and after the script finishes it will display the address of the contract deployed but also save this address to `deployments` directory on the root of this repo (except for hardhat network since this network is started and stop just for the duration of the script execution).
+1. A script to read the beacon value from the deployed contract. If selected network is not `hardhat` or `localhost` then this address must be then whitelisted by the RrpBeaconServer manager account prior to reading the beacon value.
+
+## Instructions
+
+1. Clone this repo
+1. Run the following to install the dependencies
+
+```sh
+npm install
+```
+
+Now depending on the network you are using you will need to do the following:
+
+### Default network
+
+Simplest way to test this project is by using the integrated [Hardhat network](https://hardhat.org/hardhat-network/) and not extra steps are needed in this case to run both the test and the scripts.
+
+#### Test
+
+```sh
+npx hardhat test test/BeaconReaderExample.test.js
+```
+
+#### Script
+
+```sh
+npx hardhat run scripts/deploy.js
+npx hardhat run scripts/read-beacon.js
+```
+
+### Localhost network
+
+You could also start a local ethreum node by running the following command on a separate terminal:
+
+```sh
+hardhat node
+```
+
+#### Test
+
+```sh
+npx hardhat test test/BeaconReaderExample.test.js --network localhost
+```
+
+#### Script
+
+```sh
+npx hardhat run scripts/deploy.js --network localhost
+npx hardhat run scripts/read-beacon.js --network localhost
+```
+
+### Remote networks
+
+This will require that you set some parameters in an .env file. You could rename the [.env.example](./.env.example) file in the root of this repo and replace the placeholders with valid values.
+
+For instance if you wanted to deploy the contract on [Polygon Mumbai Testnet](https://docs.polygon.technology/docs/develop/network-details/network/) you will need to set `NETWORK=polygon-mumbai` and `PROVIDER_URL=https://rpc-mumbai.matic.today`. Keep in mind that you'll need to also set the `MNEMONIC` of an account that needs to have enough funds.
+
+#### Test
+
+```sh
+npx hardhat test test/BeaconReaderExample.test.js --network polygon-mumbai
+```
+
+#### Script
+
+```sh
+npx hardhat run scripts/deploy.js --network polygon-mumbai
+npx hardhat run scripts/read-beacon.js --network polygon-mumbai
```

Comment:
Successfully got to to the following (expected) error with `rinkeby`: `error: ProviderError: execution reverted: Caller not whitelisted`, although the error was buried in a message also complaining about gas estimates:
```sh
> npx hardhat run scripts/read-beacon.js --network rinkeby
Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"name":"ProviderError","code":3,"_isProviderError":true
,"data":"0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000
01643616c6c6572206e6f742077686974656c697374656400000000000000000000"}, method="call", transaction={"from":"0x2E103c7502c1cAd0c11404F17D7C6339Db6
5626f","to":"0x6C2D6f0BC0F9b16616253fD1f1aA6AD7940DaC73","data":"0x1de7e45b6f71ad6bd1ce170abd6a64aad72131587d28696e90459e1d5d6ac8a2e889c507","ac
cessList":null}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.5.2)
    at Logger.makeError (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/logger/src.ts/index.ts:225:28)
    at Logger.throwError (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/logger/src.ts/index.ts:237:20)
    at checkError (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:80:16)
    at EthersProviderWrapper.<anonymous> (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.
ts:562:20)
    at step (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:48:23)
    at Object.throw (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:29:53)
    at rejected (/home/derek/api3/beacon-reader-example/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:21:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  reason: 'cannot estimate gas; transaction may fail or may require manual gas limit',
  code: 'UNPREDICTABLE_GAS_LIMIT',
  error: ProviderError: execution reverted: Caller not whitelisted
      at HttpProvider.request (/home/derek/api3/beacon-reader-example/node_modules/hardhat/src/internal/core/providers/http.ts:49:19)
      at HDWalletProvider.request (/home/derek/api3/beacon-reader-example/node_modules/hardhat/src/internal/core/providers/accounts.ts:187:34)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at EthersProviderWrapper.send (/home/derek/api3/beacon-reader-example/node_modules/@nomiclabs/hardhat-ethers/src/internal/ethers-provider-
wrapper.ts:13:20),
  method: 'call',
  transaction: {
    from: '0x2E103c7502c1cAd0c11404F17D7C6339Db65626f',
    to: '0x6C2D6f0BC0F9b16616253fD1f1aA6AD7940DaC73',
    data: '0x1de7e45b6f71ad6bd1ce170abd6a64aad72131587d28696e90459e1d5d6ac8a2e889c507',
    accessList: null
  }
}
``

Reply:
Yes. I'm seeing that too but I would like to get the BeaconReaderExample contract whitelisted in a network like Rinkeby and see if the error still appears when the contract is able to read the value successfully.